v2/googlecloud-to-neo4j/src/main/java/com/google/cloud/teleport/v2/neo4j/GcpToNeo4j.java

```diff
@@ -0,0 +1,339 @@
+import com.google.cloud.teleport.v2.neo4j.model.job.Source;
+import com.google.cloud.teleport.v2.neo4j.model.job.Target;
+import com.google.cloud.teleport.v2.neo4j.providers.IProvider;
+import com.google.cloud.teleport.v2.neo4j.providers.ProviderFactory;
+import com.google.cloud.teleport.v2.neo4j.transforms.GcsLogTransform;
+import com.google.cloud.teleport.v2.neo4j.transforms.Neo4jRowWriterTransform;
+import com.google.cloud.teleport.v2.neo4j.utils.BeamBlock;
+import com.google.cloud.teleport.v2.neo4j.utils.ModelUtils;
+import com.google.cloud.teleport.v2.neo4j.utils.ProcessingCoder;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.util.List;
+import org.apache.beam.repackaged.core.org.apache.commons.lang3.StringUtils;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.Wait;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Dataflow template which reads BigQuery data and writes it to Neo4j. The source data can be
+ * either a BigQuery table or an SQL query.
+ */
+public class GcpToNeo4j {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GcpToNeo4j.class);
+
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    public OptionsParams optionsParams;
+    ConnectionParams neo4jConnection;
+    JobSpec jobSpec;
+    Pipeline pipeline;
+
+    /**
+     * Main class for template.  Initializes job using run-time on pipelineOptions.
+     *
+     * @pipelineOptions framework supplied arguments
+     */
+    public GcpToNeo4j(final Neo4jFlexTemplateOptions pipelineOptions) {
+        ////////////////////////////
+        // We need to initialize pipeline in order to create context for Gs and Bq file system
+        final String jobName = pipelineOptions.getJobName() + "-" + System.currentTimeMillis();
+        pipelineOptions.setJobName(jobName);
+
+        // Set pipeline options
+        this.pipeline = Pipeline.create(pipelineOptions);
+        FileSystems.setDefaultPipelineOptions(pipelineOptions);
+        this.optionsParams = OptionsParamsMapper.fromPipelineOptions(pipelineOptions);
+        // Validate pipeline
+        processValidations(InputValidator.validateNeo4jPipelineOptions(pipelineOptions));
+
+        this.neo4jConnection = new ConnectionParams(pipelineOptions.getNeo4jConnectionUri());
+        // Validate connection
+        processValidations(InputValidator.validateNeo4jConnection(this.neo4jConnection));
+
+        this.jobSpec = JobSpecMapper.fromUri(pipelineOptions.getJobSpecUri());
+        // Validate job spec
```

Comment:
Up to you if you want to go and change this everywhere in the PR, but I'd recommend not adding comments for self-explanatory code, e.g. a comment "validate job spec" before a "validateJobSpec()" method call really only makes the code less readable :)

In general, we try to avoid adding excessive comments and comment only on non-obvious implementation details, focusing on *why* something is done and not *how* it's done. E.g. explaining *why* System.currentTimeMillis() is added to the job name about would be a good comment.

Reply:
Some of these are pseudocode stubs that I didn't remove...